### PR TITLE
add lapack dlatdf routine

### LINKED ISSUE #1651: gonum: add CARE solver
### What are you trying to do?
Add a Continuous-time algebraic Riccati equation (CARE) solver. 

If we are to follow in scipy's footsteps [source](https://github.com/scipy/scipy/blob/v1.6.3/scipy/linalg/_solvers.py#L325-L526) there are some LAPACK functions that are used that may be missing. Below are the LAPACK routines uses (dgges and dtgsen) and their dependencies. I've taken the liberty to checkmark the routines which have a file in [`gonum/lapack/gonum`](https://github.com/gonum/gonum/tree/master/lapack/gonum) (i presume that means they are implemented already):

- [ ] [dtgsen](https://github.com/scipy/scipy/blob/4ec4ab8d6ccc1cdb34b84fdcb66fde2cc0210dbf/scipy/linalg/_decomp_qz.py#L368)
- [ ] [dgges (numpylink)](https://github.com/scipy/scipy/blob/4ec4ab8d6ccc1cdb34b84fdcb66fde2cc0210dbf/scipy/linalg/_decomp_qz.py#L113) [lapack link](https://www.netlib.org/lapack/explore-html/d9/d8e/group__double_g_eeigen_ga8637d4b822e19d10327ddcb4235dc08e.html#ga8637d4b822e19d10327ddcb4235dc08e)
  - [ ] [daxpy](https://www.netlib.org/lapack/explore-html/de/da4/group__double__blas__level1_ga8f99d6a644d3396aa32db472e0cfc91c.html#ga8f99d6a644d3396aa32db472e0cfc91c)
  - [ ] [dggbal](https://www.netlib.org/lapack/explore-html/d9/d8e/group__double_g_eeigen_ga8637d4b822e19d10327ddcb4235dc08e.html#ga8637d4b822e19d10327ddcb4235dc08e)
  - [ ] [dswap](https://www.netlib.org/lapack/explore-html/de/da4/group__double__blas__level1_gaca2757ba2c3b2c6fc5d729b50345fac0.html#gaca2757ba2c3b2c6fc5d729b50345fac0)
  - [ ] [dlabad](https://www.netlib.org/lapack/explore-html/df/dd1/group___o_t_h_e_rauxiliary_ga57ac7409aedb2a380d5d0e625ffbeb7a.html#ga57ac7409aedb2a380d5d0e625ffbeb7a)
  - [ ] [dgetc2](https://www.netlib.org/lapack/explore-html/de/d39/group__double_g_eauxiliary_ga2ecd51f7842c100d6b4ecc99119c012f.html#ga2ecd51f7842c100d6b4ecc99119c012f)
  - [ ] [dgesc2](https://www.netlib.org/lapack/explore-html/de/d39/group__double_g_eauxiliary_gac6998d2ed1a92cc28708602db64f3db6.html#gac6998d2ed1a92cc28708602db64f3db6)
  - [ ] [dlatdf](https://www.netlib.org/lapack/explore-html/d8/d9b/group__double_o_t_h_e_rauxiliary_ga2744c0a54e6012d58fb4eefb75f48c65.html#ga2744c0a54e6012d58fb4eefb75f48c65)
  - [ ] [dscal](https://www.netlib.org/lapack/explore-html/de/da4/group__double__blas__level1_ga793bdd0739bbd0e0ec8655a0df08981a.html#ga793bdd0739bbd0e0ec8655a0df08981a)
  - [ ] [dtgsy2](https://www.netlib.org/lapack/explore-html/d9/df5/group__double_s_yauxiliary_gaa6de4c854b216a9afb2f6d00eaaa3b45.html#gaa6de4c854b216a9afb2f6d00eaaa3b45)
  - [ ] [dcopy](https://www.netlib.org/lapack/explore-html/de/da4/group__double__blas__level1_ga21cdaae1732dea58194c279fca30126d.html#ga21cdaae1732dea58194c279fca30126d) (surely implemented but file not present)
  - [ ] [dorgr2](https://www.netlib.org/lapack/explore-html/da/dba/group__double_o_t_h_e_rcomputational_ga92162463311fe244666f898bedcfc08f.html#ga92162463311fe244666f898bedcfc08f)
  - [ ] [dger](https://www.netlib.org/lapack/explore-html/d7/d15/group__double__blas__level2_ga458222e01b4d348e9b52b9343d52f828.html#ga458222e01b4d348e9b52b9343d52f828) 
  - [ ] [dtgexc](https://www.netlib.org/lapack/explore-html/dd/d9a/group__double_g_ecomputational_ga2510d68d70194719d570cbcfe24b3e74.html#ga2510d68d70194719d570cbcfe24b3e74)
  - [ ] [dtgex2](https://www.netlib.org/lapack/explore-html/de/d39/group__double_g_eauxiliary_gaa3c93490c68259c80285d72cb61cbd99.html#gaa3c93490c68259c80285d72cb61cbd99)

  - [ ] [dgemm](https://www.netlib.org/lapack/explore-html/d1/d54/group__double__blas__level3_gaeda3cbd99c8fb834a60a6412878226e1.html#gaeda3cbd99c8fb834a60a6412878226e1)

  - [ ] [dgghrd](https://www.netlib.org/lapack/explore-html/da/dba/group__double_o_t_h_e_rcomputational_ga14eefc9383b013ffb81cf07834c05f06.html#ga14eefc9383b013ffb81cf07834c05f06)
  - [ ] [drot](https://www.netlib.org/lapack/explore-html/de/da4/group__double__blas__level1_ga54d516b6e0497df179c9831f2554b1f9.html#ga54d516b6e0497df179c9831f2554b1f9)
  - [ ] [dhgeqz](https://www.netlib.org/lapack/explore-html/dd/d9a/group__double_g_ecomputational_ga21c89aba210fbfb8cca635e47dccbdb3.html#ga21c89aba210fbfb8cca635e47dccbdb3)
  - [ ] [dlag2](https://www.netlib.org/lapack/explore-html/d8/d9b/group__double_o_t_h_e_rauxiliary_ga4c15bef45ba06e975cc60b7d30bdd2b4.html#ga4c15bef45ba06e975cc60b7d30bdd2b4)

  - [ ] [dlagv2](https://www.netlib.org/lapack/explore-html/d8/d9b/group__double_o_t_h_e_rauxiliary_gad9f2277feb511ff205c0caf084e29109.html#gad9f2277feb511ff205c0caf084e29109)
  - [ ] [dorg2r](https://www.netlib.org/lapack/explore-html/da/dba/group__double_o_t_h_e_rcomputational_ga0598b93548008b36c26ce99b314e77e5.html#ga0598b93548008b36c26ce99b314e77e5)

  - [ ] [dgemv](https://www.netlib.org/lapack/explore-html/d7/d15/group__double__blas__level2_gadd421a107a488d524859b4a64c1901a9.html#gadd421a107a488d524859b4a64c1901a9)

  - [ ] [dtrmm](https://www.netlib.org/lapack/explore-html/d1/d54/group__double__blas__level3_gaf07edfbb2d2077687522652c9e283e1e.html#gaf07edfbb2d2077687522652c9e283e1e)
  - [ ] [dtrmv](https://www.netlib.org/lapack/explore-html/d7/d15/group__double__blas__level2_ga596c2acd9f81df6608bd5ed97e193897.html#ga596c2acd9f81df6608bd5ed97e193897)

## Already implemented routines
  - [x] [dlassq](https://www.netlib.org/lapack/explore-html/df/dd1/group___o_t_h_e_rauxiliary_ga9c824016058563e3b7189801d1de6d58.html#ga9c824016058563e3b7189801d1de6d58)
  - [x] [dlascl](https://www.netlib.org/lapack/explore-html/df/dd1/group___o_t_h_e_rauxiliary_ga7bce4c35ec5a86ee0bfdd15c476d99c8.html#ga7bce4c35ec5a86ee0bfdd15c476d99c8)
  - [x] [dlarft](https://www.netlib.org/lapack/explore-html/d8/d9b/group__double_o_t_h_e_rauxiliary_ga25cff1fc222eccac2633af51f7eff84a.html#ga25cff1fc222eccac2633af51f7eff84a)
  - [x] [dorgqr](https://www.netlib.org/lapack/explore-html/da/dba/group__double_o_t_h_e_rcomputational_ga14b45f7374dc8654073aa06879c1c459.html#ga14b45f7374dc8654073aa06879c1c459)
  - [x] [dorm2r](https://www.netlib.org/lapack/explore-html/da/dba/group__double_o_t_h_e_rcomputational_ga16f17277972002c33a44e9a7a04bb64f.html#ga16f17277972002c33a44e9a7a04bb64f)
  - [x] [dgeqrf](https://www.netlib.org/lapack/explore-html/dd/d9a/group__double_g_ecomputational_ga3766ea903391b5cf9008132f7440ec7b.html#ga3766ea903391b5cf9008132f7440ec7b)
  - [x] [dormqr](https://www.netlib.org/lapack/explore-html/da/dba/group__double_o_t_h_e_rcomputational_ga17b0765a8a0e6547bcf933979b38f0b0.html#ga17b0765a8a0e6547bcf933979b38f0b0)
  - [x] [dlacn2](https://www.netlib.org/lapack/explore-html/d8/d9b/group__double_o_t_h_e_rauxiliary_ga9b62da514b4a671acd3e3f63d018f01e.html#ga9b62da514b4a671acd3e3f63d018f01e)
  - [x] [dlacpy](https://www.netlib.org/lapack/explore-html/df/dd1/group___o_t_h_e_rauxiliary_ga7eb8731ffab2734378157c40964bf788.html#ga7eb8731ffab2734378157c40964bf788)
  - [x] [dlatrs](https://www.netlib.org/lapack/explore-html/d8/d9b/group__double_o_t_h_e_rauxiliary_gaa495c9c519906a3053f0ec56d702ef5f.html#gaa495c9c519906a3053f0ec56d702ef5f)
  - [x] [dgecon](https://www.netlib.org/lapack/explore-html/dd/d9a/group__double_g_ecomputational_ga188b8d30443d14b1a3f7f8331d87ae60.html#ga188b8d30443d14b1a3f7f8331d87ae60)
- [x] [dlaswp](https://www.netlib.org/lapack/explore-html/d8/d9b/group__double_o_t_h_e_rauxiliary_ga3ccc0cf84b0493bd9adcdc02fcff449f.html#ga3ccc0cf84b0493bd9adcdc02fcff449f)
  - [x] [dlarfb](https://www.netlib.org/lapack/explore-html/d8/d9b/group__double_o_t_h_e_rauxiliary_ga83c81583bd444e0cf021fb006cd9a5e8.html#ga83c81583bd444e0cf021fb006cd9a5e8)
  - [x] [dgeqr2](https://www.netlib.org/lapack/explore-html/dd/d9a/group__double_g_ecomputational_ga86586a8c4f1c31707724ed70e92f6f34.html#ga86586a8c4f1c31707724ed70e92f6f34)
  - [x] [dormr2](https://www.netlib.org/lapack/explore-html/da/dba/group__double_o_t_h_e_rcomputational_gadd68bb66ae7f9c7e5bc8a87154894be2.html#gadd68bb66ae7f9c7e5bc8a87154894be2)
  - [x] [dgerq2](https://www.netlib.org/lapack/explore-html/dd/d9a/group__double_g_ecomputational_ga360f5b8b756dc7d366c3decfcdf89000.html#ga360f5b8b756dc7d366c3decfcdf89000)
  - [x] [dlarf](https://www.netlib.org/lapack/explore-html/d8/d9b/group__double_o_t_h_e_rauxiliary_ga39a2b39f56ce497b218c5f48339ee712.html#ga39a2b39f56ce497b218c5f48339ee712)
- [x] [~trtrs~](https://github.com/scipy/scipy/blob/4ec4ab8d6ccc1cdb34b84fdcb66fde2cc0210dbf/scipy/linalg/basic.py#L344) [already implemented](https://github.com/gonum/gonum/blob/master/lapack/gonum/dtrtrs.go) 
  - [x] [dlaset](https://www.netlib.org/lapack/explore-html/df/dd1/group___o_t_h_e_rauxiliary_ga89e332374c7cd87e5db54bfe21550bc3.html#ga89e332374c7cd87e5db54bfe21550bc3)
  - [x] [dlarfg](https://www.netlib.org/lapack/explore-html/d8/d9b/group__double_o_t_h_e_rauxiliary_gaabb59655e820b3551af27781bd716143.html#gaabb59655e820b3551af27781bd716143)
  - [x] [dlartg](https://www.netlib.org/lapack/explore-html/df/dd1/group___o_t_h_e_rauxiliary_ga73cdfda601476bb0fdd86047dee1d357.html#ga73cdfda601476bb0fdd86047dee1d357)

Below are functions of QR decomposition so they should already be implemented
- [x] [~geqp3~](https://github.com/scipy/scipy/blob/4ec4ab8d6ccc1cdb34b84fdcb66fde2cc0210dbf/scipy/linalg/decomp_qr.py#L135)
- [x] [~geqr3~](https://github.com/scipy/scipy/blob/4ec4ab8d6ccc1cdb34b84fdcb66fde2cc0210dbf/scipy/linalg/decomp_qr.py#L139)
- [x] [~orgqr~](https://github.com/scipy/scipy/blob/4ec4ab8d6ccc1cdb34b84fdcb66fde2cc0210dbf/scipy/linalg/decomp_qr.py#L158)
- [x] [~gorgqr/gungqr~](https://github.com/scipy/scipy/blob/4ec4ab8d6ccc1cdb34b84fdcb66fde2cc0210dbf/scipy/linalg/decomp_qr.py#L161)

### Is this feature absent from the current master?
Yes

### Are you able to help contribute the feature?
I'd like to learn! I've read fortran code before but golly do I find it hard to follow. I am motivated to contribute.

### PR DESCRIPTION
**Pull request by soypat**
With respect to CARE solver to begin resolving #1651. Requires #1652 to be merged to have Dgesc2 (see lines 73-74)

### CODE REVIEW

**Comment by soypat** [comment]
PTAL

**Comment by soypat** [comment]
I will make a new PR with Dlatdf. This one has a lot of noise in the files with unwanted changes.

**Comment by vladimir-ch** [comment]
I recommend that in general you finish the implementation and its test, make sure that it passes locally and that the coverage is as complete as reasonably possible and only then submit a PR. Learning git (including rebasing) will also help keeping changes clean and under control.

**Comment by soypat** [comment]
Closing due to git rot. Will be PRing from a new repo.
